extensions/smallrye-reactive-messaging/runtime/src/main/java/io/quarkus/smallrye/reactivemessaging/runtime/QuarkusWorkerPoolRegistry.java

```diff
@@ -56,6 +114,24 @@ public <T> Uni<T> executeWork(Context currentContext, Uni<T> uni, String workerN
                 return currentContext.executeBlocking(Uni.createFrom().deferred(() -> uni), ordered);
             }
             return executionHolder.vertx().executeBlocking(uni, ordered);
+        } else if (virtualThreadConcurrency.containsKey(workerName)) {
+            Semaphore semaphore = virtualThreadConcurrency.get(workerName);
+            boolean acquired = semaphore.tryAcquire();
+            if (acquired) {
+                return runOnVirtualThread(currentContext, uni, semaphore);
+            } else {
+                if (currentContext != null) {
+                    return currentContext.executeBlocking(Uni.createFrom().deferred(() -> {
+                        semaphore.acquireUninterruptibly();
```

Comment:
Here we have lots of switches. In the worse case:

event loop -> worker -> VT

I don't have a good solution, but we should think about it. Typically, we could imagine a scheduler handling the concurrency and just scheduling the jobs when possible. It would avoid/hide the semaphore.

Reply:
I think this code is wrong even when we ignore the concurrency regulated with mutiny requests. The semaphore acquire should be done on a single-threaded executor to effectively limit the concurrency. Here we'll simply block more vert.x worker threads.

Reply:
Yes, exactly.